### ISSUE TITLE
the D and G values for updating

### ISSUE THREAD

**Issue by Mshz2** [mit-han-lab/data-efficient-gans#92]
Hi, 
would you mind to please help on how can I can find where exactly in your code, the D(T(x)), D(T(G(z))) from i, ii, and D(T(G(z)))from iii value are measured? 

![image](https://user-images.githubusercontent.com/63469924/151249986-a22ce7c2-0f7f-4b55-a7d9-b99e674d7500.png)

**Comment by zsyzzsoft** [comment]
In the code you can see `T(x)` and `T(G(z))` in e.g. [loss.py](https://github.com/mit-han-lab/data-efficient-gans/blob/master/DiffAugment-stylegan2/training/loss.py). (ii) and (iii) are not explicitly separated in the code as they are done in the optimizer level.

**Comment by Mshz2** [comment]
> In the code you can see `T(x)` and `T(G(z))` in e.g. [loss.py](https://github.com/mit-han-lab/data-efficient-gans/blob/master/DiffAugment-stylegan2/training/loss.py). (ii) and (iii) are not explicitly separated in the code as they are done in the optimizer level.

Thank you very much for your reply! So the [Loss_G](https://github.com/mit-han-lab/data-efficient-gans/blob/6858275f08f43a33026844c8c2ac4e703e8a07ba/DiffAugment-stylegan2/training/loss.py#L24) is the latest loss from Driscriminator decision on the G(z) just before be used for optimizing the Generator, right? In other words, if I wanna play around with the Generator's penalty, I only need to add values to  [Loss_G](https://github.com/mit-han-lab/data-efficient-gans/blob/6858275f08f43a33026844c8c2ac4e703e8a07ba/DiffAugment-stylegan2/training/loss.py#L24) , right?

Many Thanks!

**Comment by zsyzzsoft** [comment]
Yes, you can understand like that.

**Comment by Mshz2** [comment]
> Yes, you can understand like that.

Thanks again for your reply. One last question, I was trying to find how can I get to the Loss_G value at the training loop for every training step which leads to network update, like this in regular pytorch:
`loss.backward()`
`optimizer.step`

However, I really tried but got lost in the codes. would you please tell me where can I find this Loss_G value at each training step at the place where the training loop for each batch are happening?

I really appeaciate your efforts!

**Comment by zsyzzsoft** [comment]
This is not easy in TF1. You can change [this line](https://github.com/mit-han-lab/data-efficient-gans/blob/6858275f08f43a33026844c8c2ac4e703e8a07ba/DiffAugment-stylegan2/training/training_loop.py#L290) to e.g. `_, G_loss_val = tflib.run([G_train_op, G_loss], feed_dict)`. Alternatively, you can see `G_loss` values using `tensorboard` with the generated summary file in your experiment dir.

**Comment by Mshz2** [comment]
> This is not easy in TF1. You can change [this line](https://github.com/mit-han-lab/data-efficient-gans/blob/6858275f08f43a33026844c8c2ac4e703e8a07ba/DiffAugment-stylegan2/training/training_loop.py#L290) to e.g. `_, G_loss_val = tflib.run([G_train_op, G_loss], feed_dict)`. Alternatively, you can see `G_loss` values using `tensorboard` with the generated summary file in your experiment dir.

Thanks. I think the G_loss_val at [this line](https://github.com/mit-han-lab/data-efficient-gans/blob/6858275f08f43a33026844c8c2ac4e703e8a07ba/DiffAugment-stylegan2/training/training_loop.py#L290) is after the G already is updated by the [optimizer](https://github.com/mit-han-lab/data-efficient-gans/blob/6858275f08f43a33026844c8c2ac4e703e8a07ba/DiffAugment-stylegan2/training/training_loop.py#L232), right?

If I wanna have this G_loss value in a loop before the  [G_train_op](https://github.com/mit-han-lab/data-efficient-gans/blob/6858275f08f43a33026844c8c2ac4e703e8a07ba/DiffAugment-stylegan2/training/training_loop.py#L232) is updated, so I can modify it, in which line could I find it?

I am afraid summary file is when thinks are done and I can not change the values of G_loss for the next iteration of G net update.

**Comment by Mshz2** [comment]
> 

Then, can I instead find the mentioned loss by the [diffaugment-stylegan2-pytorch](https://github.com/mit-han-lab/data-efficient-gans/tree/master/DiffAugment-stylegan2-pytorch) in your repo? I guess since eveything, even the stylegan2 backbone, is written in pytorch

**Comment by zsyzzsoft** [comment]
Yes you can try it.
